xtask/src/main.rs

```diff
@@ -0,0 +1,328 @@
+use std::{
+    io::{BufRead, BufReader},
+    path::Path,
+    process::{ExitCode, Stdio},
+};
+
+use anyhow::{bail, ensure, Context};
+use cli::Args;
+
+use crate::{
+    cli::{Subcommand, ValidateHlslCommand, ValidateSubcommand},
+    fs::{open_file, remove_dir_all, remove_file},
+    glob::visit_files,
+    path::join_path,
+    process::{which, EasyCommand},
+    result::LogIfError,
+};
+
+mod cli;
+mod fs;
+mod glob;
+mod path;
+mod process;
+mod result;
+
+fn main() -> ExitCode {
+    env_logger::builder()
+        .filter_level(log::LevelFilter::Info)
+        .parse_default_env()
+        .format_indent(Some(0))
+        .init();
+
+    let args = Args::parse();
+
+    match run(args) {
+        Ok(()) => ExitCode::SUCCESS,
+        Err(e) => {
+            log::error!("{e:?}");
+            ExitCode::FAILURE
+        }
+    }
+}
+
+fn run(args: Args) -> anyhow::Result<()> {
+    let snapshots_base_out = join_path(["tests", "out"]);
+
+    let Args { subcommand } = args;
+
+    assert!(which("cargo").is_ok());
+
+    match subcommand {
+        Subcommand::All => {
+            EasyCommand::simple("cargo", ["fmt"]).success()?;
+            EasyCommand::simple("cargo", ["test", "--all-features", "--workspace"]).success()?;
+            EasyCommand::simple(
+                "cargo",
+                [
+                    "clippy",
+                    "--all-features",
+                    "--workspace",
+                    "--",
+                    "-D",
+                    "warnings",
+                ],
+            )
+            .success()?;
+            Ok(())
+        }
+        Subcommand::Clean => {
+            let mut found_err = false;
+            visit_files(
+                ".",
+                "*.{metal,air,metallib,vert,frag,comp,spv}",
+                &mut found_err,
+                |file, _| {
+                    remove_file(file).with_context(|| format!("failed to remove file {file:?}"))
+                },
+            );
+            ensure!(
+                !found_err,
+                "failed to clean one or more files, see above output for more details"
+            );
+            Ok(())
+        }
```

Comment:
Not sure if we need this, feels like something that was added for testing. Not a blocker though, just wanted to start the conversation.

Reply:
Let's see! Removed with 971ee01e. I've thrown a reversion into my [`ErichDonGubler/naga:xtask-clean`](https://github.com/ErichDonGubler/naga/tree/xtask-clean) branch, in case it's needed sometime soon. Otherwise, this content should always (?) be accessible in the history of this PR, if somebody actually wanted this.